server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -20,19 +29,60 @@ public class RemoteStoreRefreshListener implements ReferenceManager.RefreshListe
 
     private final Directory storeDirectory;
     private final Directory remoteDirectory;
+    // ToDo: This can be a map with metadata of the uploaded file as value of the map (GitHub #3398)
+    private final Set<String> filesUploadedToRemoteStore;
+    private static final Logger logger = LogManager.getLogger(RemoteStoreRefreshListener.class);
 
-    public RemoteStoreRefreshListener(Directory storeDirectory, Directory remoteDirectory) {
+    public RemoteStoreRefreshListener(Directory storeDirectory, Directory remoteDirectory) throws IOException {
         this.storeDirectory = storeDirectory;
         this.remoteDirectory = remoteDirectory;
+        // ToDo: Handle failures in reading list of files (GitHub #3397)
+        this.filesUploadedToRemoteStore = new HashSet<>(Arrays.asList(remoteDirectory.listAll()));
     }
 
     @Override
     public void beforeRefresh() throws IOException {
-        // ToDo Add implementation
+        // Do Nothing
     }
 
+    /**
+     * Upload new segment files created as part of the last refresh to the remote segment store.
+     * The method also deletes segment files from remote store which are not part of local filesystem.
+     * @param didRefresh true if the refresh opened a new reference
+     * @throws IOException in case of I/O error in reading list of local files
+     */
     @Override
     public void afterRefresh(boolean didRefresh) throws IOException {
-        // ToDo Add implementation
+        if (didRefresh) {
+            Set<String> localFiles = Arrays.stream(storeDirectory.listAll()).collect(Collectors.toSet());
+            localFiles.stream().filter(file -> !filesUploadedToRemoteStore.contains(file)).forEach(file -> {
+                try {
+                    remoteDirectory.copyFrom(storeDirectory, file, file, IOContext.DEFAULT);
+                    filesUploadedToRemoteStore.add(file);
+                } catch (NoSuchFileException e) {
+                    logger.info(
```

Comment:
Could a possible segment merge with a concurrency bug manifest in files to be removed?